orttraining/orttraining/python/training/ortmodule/experimental/hierarchical_ortmodule/_hierarchical_ortmodule.py

```diff
@@ -0,0 +1,154 @@
+        # The function "record_args" records the inputs for each nn.Module,
+        # and later we will try exporting those nn.Module's with their recorded
+        # inputs.
+        def recursive_hook(module):
+            handle_pool.append(module.register_forward_pre_hook(record_args))
+            for name, sub in module._modules.items():
+                if isinstance(sub, torch.nn.ModuleList):
+                    for name1, sub1 in sub._modules.items():
+                        recursive_hook(sub1)
+                else:
+                    recursive_hook(sub)
+
+        exportable_list = {}
+
+        # Fill "exportable_list". exportable_list[module] = True means
+        # "module" can be wrapped as ORTModule. Otherwise, "module" is
+        # not exportable to ONNX.
+        def check_exportable(module):
+            sub_dict = module._modules
+            if not sub_dict:
+                # No sub-module exists, so this module is a leaf
+                # module in overall model hierarchy.
+                exportable = True
+                # Check if this leaf module is exportable.
+                for args in module_arg_pool[module]:
+                    try:
+                        with tempfile.NamedTemporaryFile(prefix='sub-module') as temp:
+                            torch.onnx.export(
+                                module, args, temp, opset_version=9)
+                    except Exception as e:
+                        exportable = False
+
+                exportable_list[module] = exportable
+                return exportable
+            else:
+                sub_exportable = True
+                for name, sub_module in sub_dict.items():
+                    if isinstance(sub_module, torch.nn.ModuleList):
+                        for name1, sub_module1 in sub_module._modules.items():
+                            sub_exportable1 = check_exportable(sub_module1)
+                            sub_exportable = sub_exportable and sub_exportable1
+                    else:
+                        sub_exportable1 = check_exportable(sub_module)
+                        sub_exportable = sub_exportable and sub_exportable1
+
+                if sub_exportable is False:
+                    # At least one existing sub-module is not exportable,
+                    # so is the entire module.
+                    exportable_list[module] = sub_exportable
+                    return sub_exportable
+                else:
+                    # Now, we know all sub-modules are exportable, so
+                    # we are going to check if the composition of them
+                    # is still exportable at this module level.
+                    module_exportable = True
+                    for args in module_arg_pool[module]:
+                        try:
+                            with tempfile.NamedTemporaryFile(prefix='sub-module') as temp:
+                                torch.onnx.export(
```

Comment:
I was wondering should we align this export parameters with the ones we used for common ORTModule export. One thing among them is, for PythonOp, we need enable_grad as the context to make sure it has the correct requires_grad for execution.

Reply:
I think we could improve this step by step, not immediately needed in this PR. Up to you.

Reply:
That makes sense. Any bug fixed on `ORTModule` would still affect `HierarchicalModule` until the code is duplicated here

Reply:
Just change it to
```
                                torch.onnx.export(
                                    module, args, temp, opset_version=ONNX_OPSET_VERSION,
                                    do_constant_folding=False, export_params=False,
                                    keep_initializers_as_inputs=True,
                                    training=torch.onnx.TrainingMode.TRAINING)
```